packages/client/tiptap/extensions/database/data.ts

```diff
@@ -0,0 +1,120 @@
+import * as Y from 'yjs'
+import {DataType} from './types'
+
+export type ColumnId = string
+export type RowId = string
+export type ColumnMeta = {
+  name: string
+  type: DataType
+}
+
+const cloneColumnMeta = (meta: ColumnMeta): ColumnMeta => {
+  return {
+    name: meta.name,
+    type: meta.type
+  }
+}
+
+const generateId = (doc: Y.Doc) => {
+  const id = Math.random().toString(36).substring(2, 9)
+  for (const column of doc.getArray<ColumnId>('columns')) {
+    if (id === column) {
+      return generateId(doc)
+    }
+  }
+  for (const row of doc.getArray<RowId>('rows')) {
+    if (id === row) {
+      return generateId(doc)
+    }
+  }
+  return id
+}
+
+export const changeColumn = (doc: Y.Doc, columnId: ColumnId, newMeta: ColumnMeta) => {
+  const columnMeta = doc.getMap<ColumnMeta>('columnMeta')
+  columnMeta.set(columnId, newMeta)
+}
+
+// Needs to be added to a column!
+const createDanglingColumn = (doc: Y.Doc) => {
```

Comment:
-1 for this and all the other functions below, we need to wrap all the writes in a `yDoc.transact` because an abrupt exit could corrupt the data